presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchange.java

```diff
@@ -275,23 +289,106 @@ private static void checkNotHoldsLock(Object lock)
                 PartitioningHandle partitioning,
                 int defaultConcurrency,
-                List<Type> types,
-                List<Integer> partitionChannels,
-                Optional<Integer> partitionHashChannel,
+                List<Type> inputTypes,
+                List<Integer> partitioningChannels,
+                Optional<Integer> hashChannel,
                 PipelineExecutionStrategy exchangeSourcePipelineExecutionStrategy,
                 DataSize maxBufferedBytes)
+        {
+            int outputPartitionCount = computeBufferCount(partitioning, defaultConcurrency, partitioningChannels);
+            List<Type> partitioningChannelTypes = partitioningChannels.stream()
+                    .map(inputTypes::get)
+                    .collect(toImmutableList());
+            Optional<PartitionFunction> partitionFunction = getPartitionFunction(
+                    partitioningProviderManager,
+                    session,
+                    partitioning,
+                    outputPartitionCount,
+                    partitioningChannelTypes,
+                    hashChannel.isPresent());
+            return new LocalExchangeFactory(
+                    partitioning,
+                    partitionFunction,
+                    partitioningChannels,
+                    hashChannel,
+                    exchangeSourcePipelineExecutionStrategy,
+                    maxBufferedBytes,
+                    outputPartitionCount);
+        }
+
+        public static Optional<PartitionFunction> getPartitionFunction(
+                PartitioningProviderManager partitioningProviderManager,
+                Session session,
+                PartitioningHandle partitioning,
+                int partitionCount,
+                List<Type> partitioningChannelTypes,
+                boolean isHashPrecomputed)
+        {
+            if (partitioning.equals(SINGLE_DISTRIBUTION) ||
+                    partitioning.equals(FIXED_BROADCAST_DISTRIBUTION) ||
+                    partitioning.equals(FIXED_ARBITRARY_DISTRIBUTION) ||
+                    partitioning.equals(FIXED_PASSTHROUGH_DISTRIBUTION)) {
+                // not partitioned exchange
+                return Optional.empty();
+            }
+
+            if (partitioning.getConnectorHandle() instanceof SystemPartitioningHandle) {
+                HashGenerator hashGenerator;
+                if (isHashPrecomputed) {
+                    hashGenerator = new PrecomputedHashGenerator(0);
+                }
+                else {
+                    hashGenerator = new InterpretedHashGenerator(partitioningChannelTypes, IntStream.range(0, partitioningChannelTypes.size()).toArray());
+                }
+                return Optional.of(new LocalPartitionGenerator(hashGenerator, partitionCount));
+            }
+
+            ConnectorNodePartitioningProvider partitioningProvider = partitioningProviderManager.getPartitioningProvider(partitioning.getConnectorId().get());
```

Comment:
I feel this part of code looks very familiar, is there any way to reusing `NodePartitioningManager#getPartitionFunction` instead of replicating this quite complicated partition function generation algorithm ?

Reply:
Although there many similarities, there are two important differences:

1. System hash function is different. System hash function for local exchange is suppose to re-hash the hash to reuse entropy from different hash bits to minimize collisions. So the part related to the `System` partitioning is different.
2. `NodePartitioningManager#getPartitionFunction` requires `PartitioningScheme` with the `Optional<int[]> bucketToPartition` set. This function computes `bucketToPartition` based on the number of partitions providing a cleaner signature (`partitioning` + `partitionCount` vs `PartitioningScheme` with implicit assumption that the `bucketToPartition` is set.)